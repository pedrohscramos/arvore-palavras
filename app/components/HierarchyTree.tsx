import React, { useState } from 'react';
import WordInput from './WordInput';

type Node = {
  [key: string]: any;
};

const HierarchyTree: React.FC = () => {
  const [tree, setTree] = useState<Node>({});

  const addNode = (parentPath: string[], nodeName: string) => {
    const newTree = { ...tree };
    let currentNode: Node = newTree;

    // Navegar até o nó pai usando o caminho fornecido
    parentPath.forEach((key) => {
      if (!currentNode[key]) {
        currentNode[key] = {};
      }
      currentNode = currentNode[key];
    });

    // Verificar se é uma lista de palavras (último nível) ou uma subcategoria (nível intermediário)
    if (Array.isArray(currentNode)) {
      currentNode.push(nodeName);
    } else {
      currentNode[nodeName] = [];
    }

    setTree(newTree);
  };

  const renderTree = (node: Node, level = 0) => {
    return Object.keys(node).map((key, index) => (
      <div key={index} style={{ marginLeft: `${level * 20}px` }}>
        {key}
        {Array.isArray(node[key]) ? (
          <ul>
            {node[key].map((item: string, idx: number) => (
              <li key={idx}>{item}</li>
            ))}
          </ul>
        ) : (
          renderTree(node[key], level + 1)
        )}
      </div>
    ));
  };

  const saveHierarchy = () => {
    const json = JSON.stringify(tree, null, 2);
    const blob = new Blob([json], { type: 'application/json' });
    const link = document.createElement('a');
    link.href = URL.createObjectURL(blob);
    link.download = 'hierarchy.json';
    link.click();
  };

  return (
    <div>
      <h1>Árvore de palavras</h1>
      <WordInput onAddNode={addNode} />
      <div>{renderTree(tree)}</div>
      <button onClick={saveHierarchy}>Salvar Árvore</button>
    </div>
  );
};

export default HierarchyTree;
