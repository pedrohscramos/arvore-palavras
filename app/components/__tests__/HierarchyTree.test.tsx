import '@testing-library/jest-dom';
import { fireEvent, render, screen } from '@testing-library/react';
import HierarchyTree from '../HierarchyTree';

describe('HierarchyTree Component', () => {
  it('should render hierarchy builder elements', () => {
    render(<HierarchyTree />);

    expect(screen.getByText('Árvore de palavras')).toBeInTheDocument();
    expect(screen.getByText('Salvar Árvore')).toBeInTheDocument();
  });

  it('should add nodes and render the hierarchy correctly', () => {
    render(<HierarchyTree />);

    // Adiciona uma palavra no nível raiz
    fireEvent.change(screen.getByPlaceholderText('Adicione uma palavra'), { target: { value: 'Animais' } });
    fireEvent.change(screen.getByPlaceholderText('Caminho (ex: Animais.Mamíferos)'), { target: { value: '' } });
    fireEvent.click(screen.getByText('Adicionar Palavra'));

    // Verifica se a palavra "Animais" aparece no DOM
    expect(screen.getByText('Animais')).toBeInTheDocument();

    // Adiciona uma subcategoria
    fireEvent.change(screen.getByPlaceholderText('Adicione uma palavra'), { target: { value: 'Mamíferos' } });
    fireEvent.change(screen.getByPlaceholderText('Caminho (ex: Animais.Mamíferos)'), { target: { value: 'Animais' } });
    fireEvent.click(screen.getByText('Adicionar Palavra'));

    // Verifica se a subcategoria "Mamíferos" aparece aninhada em "Animais"
    expect(screen.getByText('Mamíferos')).toBeInTheDocument();
  });

  it('should download the hierarchy as a JSON file when clicking "Salvar Árvore"', () => {
    const { container } = render(<HierarchyTree />);

    global.URL.createObjectURL = jest.fn();
    // Simula a adição de uma palavra
    fireEvent.change(screen.getByPlaceholderText('Adicione uma palavra'), { target: { value: 'Animais' } });
    fireEvent.click(screen.getByText('Adicionar Palavra'));

    // Simula o clique no botão "Salvar Árvore"
    const saveButton = screen.getByText('Salvar Árvore');
    const createElementSpy = jest.spyOn(document, 'createElement');
    fireEvent.click(saveButton);

    // Verifica se o arquivo foi gerado e o link de download foi criado
    expect(createElementSpy).toHaveBeenCalledWith('a');
    expect(container).toMatchSnapshot();
  });
});
