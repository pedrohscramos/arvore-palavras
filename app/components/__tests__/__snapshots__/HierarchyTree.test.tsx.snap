// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HierarchyTree Component should download the hierarchy as a JSON file when clicking "Salvar Árvore" 1`] = `
<div>
  <div>
    <h1>
      Árvore de palavras
    </h1>
    <div>
      <input
        placeholder="Adicione uma palavra"
        type="text"
        value=""
      />
      <input
        placeholder="Caminho (ex: Animais.Mamíferos)"
        type="text"
        value=""
      />
      <button>
        Adicionar Palavra
      </button>
    </div>
    <div>
      <div
        style="margin-left: 0px;"
      >
        Animais
        <ul />
      </div>
    </div>
    <button>
      Salvar Árvore
    </button>
  </div>
</div>
`;
